cpp/src/arrow/dataset/dataset.h

```diff
@@ -37,17 +38,91 @@ namespace dataset {
 
 using RecordBatchGenerator = std::function<Future<std::shared_ptr<RecordBatch>>()>;
 
+/// \brief Description of a column to scan
+struct FragmentSelectionColumn {
+  /// \brief The path to the column to load
+  FieldPath path;
+  /// \brief The type of the column in the dataset schema
+  ///
+  /// A format may choose to ignore this field completely.  For example, when
+  /// reading from IPC the reader can just return the column in the data type
+  /// that is stored on disk.  There is no point in doing anything special.
+  ///
+  /// However, some formats may be capable of casting on the fly.  For example,
+  /// when reading from CSV, if we know the target type of the column, we can
+  /// convert from string to the target type as we read.
+  DataType* requested_type;
+  /// \brief The index in the output selection of this column
+  int selection_index;
+};
+/// \brief Instructions for scanning a particular fragment
+///
+/// The fragment scan request is dervied from ScanV2Options.  The main
+/// difference is that the scan options are based on the dataset schema
+/// while the fragment request is based on the fragment schema.
+struct FragmentScanRequest {
+  /// \brief A row filter
+  ///
+  /// The filter expression should be written against the fragment schema.
+  ///
+  /// \see ScanV2Options for details on how this filter should be applied
+  compute::Expression filter = compute::literal(true);
+
+  /// \brief The columns to scan
+  ///
+  /// These indices refer to the fragment schema
+  ///
+  /// Note: This is NOT a simple list of top-level column indices.
+  /// For more details \see ScanV2Options
+  ///
+  /// If possible a fragment should only read from disk the data needed
+  /// to satisfy these columns.  If a format cannot partially read a nested
+  /// column (e.g. JSON) then it must apply the column selection (in memory)
+  /// before returning the scanned batch.
+  std::vector<FragmentSelectionColumn> columns;
+  /// \brief Options specific to the format being scanned
+  FragmentScanOptions* format_scan_options;
+};
+
+class FragmentScanner {
+ public:
+  /// This instance will only be destroyed after all ongoing scan futures
+  /// have been completed.
+  ///
+  /// This means any callbacks created as part of the scan can safely
+  /// capture `this`
+  virtual ~FragmentScanner() = default;
+  /// \brief Scan a batch of data from the file
+  /// \param batch_number The index of the batch to read
+  virtual Future<std::shared_ptr<RecordBatch>> ScanBatch(int batch_number) = 0;
+  /// \brief Calculate an estimate of how many data bytes the given batch will represent
+  ///
+  /// "Data bytes" should be the total size of all the buffers once the data has been
+  /// decoded into the Arrow format.
+  virtual int64_t EstimatedDataBytes(int batch_number) = 0;
+  /// \brief The number of batches in the fragment to scan
+  virtual int NumBatches() = 0;
```

Comment:
How would something like CSV implement this?

Reply:
Should NumBatches and EstimatedDataBytes be `const`?

Reply:
There is no way to know the number of batches, no. And I'd like to avoid reimplementing parallel scans for Flight/ADBC data sources…

Reply:
Hmm, and I guess `(fragment_index, batch_index, is_last)` pushes too much complexity into the ordered execution bits?

Reply:
Actually I guess I'm curious how this will work with iterator-based sources in general. Flight/ADBC would give you a set of iterators, where there is no parallelism within an iterator. And if you wanted to support Arrow stream files for some reason, that'd have the same structure. 

On the other hand, Flight/ADBC have a fixed number of fragments up front.

Maybe a separate source node for those sources might make sense, with an optional 'merge' step if you want to use ordered execution with them? It would be kind of a shame to force them outside of the Dataset framework entirely, IMO, but having two different interfaces would also be annoying. I'm kind of just typing out things on the fly here…

Reply:
I think the worst case would be if the fragments are large and there is no parallelism within an iterator and the user wants ordered execution.  This will be a problem regardless of approach.  We will need to queue a large number of batches in memory for any kind of resequencing and, as a result, will probably fallback to serial execution.

> Hmm, and I guess (fragment_index, batch_index, is_last) pushes too much complexity into the ordered execution bits?

It might not be too bad.  I'll take another look.